[DNM] Revert "IRGen: Use llvm.compiler.used instead of llvm.used on ELF"

kateinoigakukun opened the pull request:
This reverts commit 2bbea54dc0bf95d2622f71ecc3e100eff3a19183 to see what multiple sections for each global broke.

The use of `llvm.compiler.used` as a replacement for `llvm.used` changes the semantics of symbol requirements at link time, and it loses `SHF_GNU_RETAIN` for generated `swift5` metadata sections.

I'm not sure why multiple same name sections were problem here 🤔

keith committed:
Mark start/stop sections as SHF_GNU_RETAIN

Fixes https://github.com/apple/swift-package-manager/issues/5698

kateinoigakukun committed:
Revert "IRGen: Use llvm.compiler.used instead of llvm.used on ELF"

This reverts commit 2bbea54dc0bf95d2622f71ecc3e100eff3a19183.

kateinoigakukun commented:
@swift-ci Please test

aschwaighofer commented:
The problem is/was that you end up with multiple section of the same name. 

The runtime (CC @mikeash) and tools look for said section and (presumably) can't deal with the presence of multiple same named sections in one binary on linux and therefore you see failures.

Speculating ..., to address this, maybe the runtime could be modified? I am assuming that this means something like: if you have seen one section of  (e.g) `swift5_protocol_conformances` continue looking at potentially other sections of the same name.

compnerd commented:
The section data should be coalesced, so it seems like we might be missing something.

aschwaighofer commented:
The fact that they are not coalesced seems to be intentional feature of SHF_GNU_RETAIN. https://reviews.llvm.org/D97448

> For front ends which do not expect to see multiple sections of the same name,
consider emitting @llvm.compiler.used instead of @llvm.used.

kateinoigakukun commented:
IIUC, multiple same-name sections are OK in object files, but the problem here is linker does not concatenate those same-name sections appeared in a single object file, but just pick a first one. I will have a closer look again.

kateinoigakukun commented:
Okay, the problem was that:
1. gold linker creates separate sections for same name input sections with different section flags. https://sourceware.org/git/?p=binutils-gdb.git;a=blob;f=gold/layout.cc;h=b43ae841a6cc90689d6e62244ce6fa83f9be885b;hb=HEAD#l845
2. After reverting 9ce1baf3f3dd42567ae53409d277be37b23a2ce1, each `swift5` section emitted by compiler has `SHF_GNU_RETAIN`, but `swiftrt.o`'s sections didn't.
3. `swiftrt.o` is placed before Swift object files in gold linker command arguments, so gold creates encapsulation symbols (start/stop) for the swiftrt.o's empty sections.

The 1. behavior is inconsistent between gold and lld (lld merges those sections into one), so it was not a problem with lld.

I applied 0e410da168b014733142c07916fb899091de96fe to add `SHF_GNU_RETAIN` flags to the `swiftrt.o`'s empty sections to have consistent section flags with compiler-emitted object files.

I tested with the following env var and it works on my end.
```
SWIFT_DRIVER_TEST_OPTIONS=" -Xlinker --gc-sections -use-ld=lld"
SWIFT_DRIVER_TEST_OPTIONS=" -Xlinker --gc-sections -use-ld=gold"
```

kateinoigakukun commented:
@swift-ci Please test Linux platform

kateinoigakukun committed:
Stop using `SHF_GNU_RETAIN` for some reflection sections

To have consistent section flags between compiler emitted object files
and swiftrt.o.

kateinoigakukun commented:
I saw some of the reflection tests failed because some of reflection sections like `swift5_typeref` and `swift5_reflstr` are not flagged with `SHF_GNU_RETAIN`, but `swiftrt.o` has empty sections with the flag. 05e2fc1fb47260d0d3cf947b3483dd794987fb17 fixes part of the problem by removing the flag for the reflection sections.

However, some of reflection sections like `swift5_fieldmd` are usually retained but not retained when using `-reflection-metadata-for-debugger-only` or `-disable-reflection-metadata`.

https://github.com/apple/swift/blob/aa5b505014b4f622d5f1a9846f1a48d3ba7118fc/lib/IRGen/GenReflection.cpp#L724-L729

This conditional section flag decision makes it difficult to have consistent section flags between object files, and it results in multiple reflection sections in the final image.

I think one of the solutions would be adding retained versions of those conditionally retained sections `swift5_fieldmd_retain`, but it adds 5 more `swift5` sections. What do you think about this solution? and do you have any other good idea?

kateinoigakukun commented:
I saw some of the reflection tests failed because some of reflection sections like `swift5_typeref` and `swift5_reflstr` are not flagged with `SHF_GNU_RETAIN`, but `swiftrt.o` has empty sections with the flag. 05e2fc1fb47260d0d3cf947b3483dd794987fb17 fixes part of the problem by removing the flag for the reflection sections.

However, some of reflection sections like `swift5_fieldmd` are usually retained but not retained when using `-reflection-metadata-for-debugger-only` or `-disable-reflection-metadata`.

https://github.com/apple/swift/blob/aa5b505014b4f622d5f1a9846f1a48d3ba7118fc/lib/IRGen/GenReflection.cpp#L724-L729

This conditional section flag decision makes it difficult to have consistent section flags between object files, and it results in multiple reflection sections in the final image.

I think one of the solutions would be adding retained versions of those conditionally retained sections `swift5_fieldmd_retain`, but it adds 5 more `swift5` sections. 

What do you think about this solution? and do you have any other good ideas? @compnerd @aschwaighofer

compnerd commented:
Could we initialize the section appropriately so that the attributes are applied to that section into which we emit the symbols? We can change the behaviour conditionally based on the object file format that we are emitting into as well if that is necessary.

kateinoigakukun commented:
@compnerd Do you mean not to have `SHF_GNU_RETAIN` section flag for those reflection metadata sections consistently and control retainability by something else? Or do you mean always to mark those reflection metadata retained on ELF object files and not supporting reflection metadata stripping on ELF?

compnerd commented:
I was thinking more the latter. We can emit them if reflection is enabled and not strip them.

al45tair commented:
> Or do you mean always to mark those reflection metadata as "retained" on ELF object files and not supporting reflection metadata stripping on ELF?

I think this is probably the way forward in the short term. WDYT @aschwaighofer?

al45tair commented:
(I could do with having this resolved for my static Linux SDK work — see #71841, which is one of the changes I had on my branch, and which, of course, doesn't work in CI because of this problem.)

al45tair commented:
FWIW, I just filed a bug against `gold`, here: https://sourceware.org/bugzilla/show_bug.cgi?id=31415

keith commented:
> This is then compounded by the fact that `gold` does not coalesce sections whose flags differ

My assumption was that we could make it so the flags never differ for these sections. Presumably by making a similar change to https://github.com/apple/swift/pull/71841 wherever that is in the code, is that not the case?

al45tair commented:
> My assumption was that we could make it so the flags never differ for these sections. Presumably by making a similar change to #71841 wherever that is in the code, is that not the case?

There is [the issue @kateinoigakukun mentions above](https://github.com/apple/swift/pull/71373#issuecomment-1931434000), where the reflection metadata is marked either used or not used depending on a flag, which complicates matters slightly.

One other option that nobody has mentioned so far is for the runtime to support _two_ sections (with different names, hence different `__start` and `__stop` symbols) for every one where this is a problem, then have the compiler put things into one or other of them depending on the setting of the flag. That's doable and might be worth exploring.

kubamracek commented:
> > Or do you mean always to mark those reflection metadata as "retained" on ELF object files and not supporting reflection metadata stripping on ELF?

> I think this is probably the way forward in the short term. WDYT @aschwaighofer @mikeash? Also, I guess @kubamracek might have some opinions on this?

I'm okay if we for now declare that reflection metadata stripping on ELF won't work. It's only needed for the -experimental-hermetic-seal-at-link / WME / VFE features, which don't (sadly) work properly on ELF today anyway.

kubamracek commented:
Not sure which exact part you're asking about. The -experimental-hermetic-seal-at-link feature can work with reflection metadata being completely off, but that regresses debugging because without reflection metadata, LLDB doesn't know how to inspect variables and understand their types. Runtime execution will still work, though.

keith commented:
Sorry I was just talking about the section presence part. If the section is there, but empty, I assume that would be ok?

al45tair commented:
I've been looking at this some more. The _runtime_ only cares about the section if it's retained, so `swiftrt.o` only needs to include declarations for the retained section. We don't need to turn off start-stop GC. We also don't want to use `@llvm.compiler.used`, because we really do want the retain flag.

The reflection code is where we might care about the unretained sections, so `readELFSections()` in `ReflectionContext.h` needs changing to look for both retained and unretained sections. The easiest solution here seems to be to make the compiler generate a different section name for the non-retained sections, which will mean that all the linkers will behave the same way if you link retained and unretained reflection metadata into the same program. Then we can just make `readELFSections()` add an extra `ReflectionInfo` if it finds unretained sections.

That _should_ fix everything I think. I've got a changeset for all of this and I'm currently trying the test suite with
```
SWIFT_DRIVER_TEST_OPTIONS=" -Xlinker --gc-sections -use-ld=lld"
SWIFT_DRIVER_TEST_OPTIONS=" -Xlinker --gc-sections -use-ld=gold"
```
as @kateinoigakukun did previously to make sure I haven't broken anything.

al45tair commented:
I have a new PR that deals with all of this: #72061.

kateinoigakukun commented:
Great!
